iOS Simulator and iOS platform support

gpx1000 opened the pull request:
## Description

Add support for iOS

Fixes #627 

## General Checklist:

Please ensure the following points are checked:

- [ ] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [ ] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [ ] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [ ] I have commented any added functions (in line with Doxygen)
- [ ] I have commented any code that could be hard to understand
- [ ] My changes do not add any new compiler warnings
- [ ] My changes do not add any new validation layer errors or warnings
- [ ] I have used existing framework/helper functions where possible
- [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [ ] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [ ] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [ ] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [ ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [ ] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

gpx1000 committed:
ios now builds.  Had to update a lot of our dependencies to get this far.  Need to do documentation and testing.  Also need to probably port the Android sample selection menu.

gpx1000 committed:
try to fix the clang format and copyright.

gpx1000 committed:
forgot these two

gpx1000 committed:
one more clang format fix.

gpx1000 committed:
fix copyright headers and doxygen

gpx1000 committed:
fix OSX builds again

gpx1000 committed:
Odd that this wasn't fixed for the clang-format?!?

gpx1000 committed:
Merge branch 'main' into ios_platform

# Conflicts:
#	app/main.cpp
#	framework/platform/unix/unix_platform.cpp

gpx1000 committed:
update from main and fix not building.  Need to add instructions for how to build and run.

gpx1000 committed:
This builds and runs now.

gpx1000 committed:
add build guide for iOS

gpx1000 committed:
fix copyright and start the tidy fixes

gpx1000 committed:
add ios to the CI

gpx1000 committed:
add ios to the CI

gpx1000 commented:
NB: this fails currently on the VolkInitialize().

gpx1000 committed:
fix the formating.

gpx1000 committed:
fix the CI.  Add system MoltenVK framework incase not using loader.

gpx1000 committed:
Try installing vulkan first for the CI.

gpx1000 committed:
fix the clang format.

gpx1000 committed:
fix hdiutil not being found in CI

gpx1000 committed:
fix clang-format.

gpx1000 committed:
Try to properly detach VulkanSDK installer and run the environment setup script so hopefully cmake can find it in CI.

gpx1000 committed:
fix the clang-format

gpx1000 committed:
attempt to cache the VulkanSDK install

gpx1000 committed:
Try using the Vulkan package instead of MoltenVK package.

gpx1000 committed:
give Vulkan a hint for the CI to find the SDK in the home directory for Ios.

gpx1000 committed:
Cleanup the cmake warning in a proper way.  I couldn't get the moltenvk framework to link from Vulkan::MoltenVK as find_vulkan doesn't work well in ios for whatever reason.  So doing it by hand.

gpx1000 committed:
enforce linking MoltenVK by hand.

gpx1000 committed:
link MoltenVK to the main program.

Use OSX 13

gpx1000 committed:
switch to using something other than xcode so it can be cached with ccache.

gpx1000 committed:
revert the compile on other than xcode

tomadamatkinson commented on docs/build.md:
````diff
@@ -269,6 +269,57 @@ cmake --build build/mac --config Release --target vulkan_samples -j4
+
+```
+cmake --build build/ios --config Release --target 
+vulkan_samples -j4 -sdk iphonesimulator -allowProvisioningUpdates
````
Does this need to be on a single line?

tomadamatkinson commented on bldsys/cmake/global_options.cmake:
```diff
@@ -32,12 +32,9 @@ else()
-    cmake_minimum_required(VERSION 3.19)
-    find_package(Vulkan 1.3...<2.0)
-    if(Vulkan_FOUND)
+    cmake_minimum_required(VERSION 3.24)
```
Can we require a CMake version smaller than 3.24?

tomadamatkinson replied:
I have updated now from 3.22 to 3.27 on my Macbook. But it is always best to use the minimum version. Happy to keep this change if 3.24 is the new required version

gpx1000 replied:
TBH, with the new VulkanSDK coming out where we use xcframework, we're going to need to bump it to 3.28  But we can work around that if we need to.

tomadamatkinson commented on docs/build.md:
````diff
@@ -269,6 +269,57 @@ cmake --build build/mac --config Release --target vulkan_samples -j4
+vulkan_samples -j4 -sdk iphonesimulator -allowProvisioningUpdates
+```
+
+NB: The above creates a bundle.  It is possible to build 
````
Is it worth adding Xcode instructions?

For me this was the following with my iphone plugged in

On the Iphone
1. Go the Settings app, and navigate to Privacy & Security > Developer Mode
2. Restart when prompted

On the macbook
1. Open Xcode -> Open the Xcode project from build/ios
3. Set the target to vulkan_samples
4. Follow code signing steps https://developer.apple.com/documentation/xcode/running-your-app-in-simulator-or-on-a-device
5. Set the device to a simulator (default) or the plugged in iphone
6. Click Play

We could also make an issue to follow up with these docs and do it at a later date as these would likely require screenshot examples

tomadamatkinson replied:
There are a lot of code signing steps but I eventually got the application to run

gpx1000 replied:
I have now gotten this to work without XCode needing to be open.  However, now, you can simply open the project in XCode and hitting the play button; it'll simply work no need for the extra work.

tomadamatkinson commented on framework/platform/ios/ios_window.mm:
```diff
@@ -0,0 +1,84 @@
+
+float IosWindow::get_dpi_factor() const
+{
+	//	return AConfiguration_getDensity(platform->get_android_app()->config) / static_cast<float>(ACONFIGURATION_DENSITY_MEDIUM);
```
remove unused code

tomadamatkinson commented on framework/platform/ios/ios_window.mm:
```diff
@@ -0,0 +1,84 @@
+
+void IosWindow::close()
+{
+	//	GameActivity_finish(platform->get_activity());
```
remove unused code

tomadamatkinson commented on framework/platform/ios/ios_window.mm:
```diff
@@ -0,0 +1,84 @@
+
+void IosWindow::process_events()
+{
+	//	process_android_events(platform->get_android_app());
```
remove unused code

tomadamatkinson commented on app/CMakeLists.txt:
```diff
@@ -80,4 +80,19 @@ if(MSVC)
+    set(CMAKE_MACOSX_BUNDLE YES)
+    set(MACOSX_BUNDLE_GUI_IDENTIFIER com.khronos.vulkan_samples)
+    set_target_properties(${PROJECT_NAME} PROPERTIES
+            BUNDLE_IDENTIFIER com.khronos.vulkan_samples
```
Code signing doesn't work with `com.khronos.vulkan_samples` but did work with `com.khronos.VulkanSamples`

tomadamatkinson reviewed (changes requested):
The application starts but Vulkan initialization fails on the device and in the simulator. Did you need to perform any extra steps to get Vulkan to work?

gpx1000 committed:
Merge branch 'main' into ios_platform

# Conflicts:
#	docs/build.md
#	framework/CMakeLists.txt
#	framework/core/sampler_core.cpp
#	framework/gui.cpp
#	framework/gui.h
#	framework/hpp_gui.cpp
#	framework/hpp_gui.h
#	framework/scene_graph/components/image/astc.cpp
#	framework/scene_graph/components/image/astc.h
#	samples/api/hpp_terrain_tessellation/hpp_terrain_tessellation.cpp
#	third_party/CMakeLists.txt

gpx1000 committed:
iOS updated and works against triangle.  Need to investigate updating hpp vulkan to allow for the new framework.

gpx1000 committed:
Fix for other samples and vulkan hpp.

gpx1000 committed:
remove excess changes.

gpx1000 committed:
address requested changes.

gpx1000 committed:
Merge branch 'main' into ios_platform

gpx1000 committed:
remove extra define.

gpx1000 committed:
upgrade imgui to get past the build problem

gpx1000 committed:
upgrade imgui to get past the build problem

gpx1000 committed:
calm the CI so it can validate the build.

gpx1000 committed:
add in missing copyrights

gpx1000 committed:
prevent build_ios task from running until the new SDK is released.

gpx1000 committed:
All samples now work if they work in MacOS they work in iOS.

gpx1000 committed:
cleanup the CMake to show only frameworks are needed.

gpx1000 committed:
add basic run documentation.

gpx1000 committed:
FindVulkan.cmake didn't make it into the SDK, so package it here.

gpx1000 committed:
1.3.280 Vulkan SDK has been released and is necessary for iOS support.  Add CI runner back in.

gpx1000 committed:
sdk names itself differently?

gpx1000 committed:
fix a crashbug

gpx1000 committed:
VULKAN_SDK as set in setup_env.sh points to the macOS folder.

gpx1000 committed:
work around bug in CMake prior to 3.29

gpx1000 committed:
fix the copyright header check

gpx1000 committed:
figure out what github CI is doing.

gpx1000 committed:
Tell VulkanSDK to install the iOS support components.

gpx1000 committed:
see if this fixes ASTC trying to link a x86-64 library when it's building for iOS.

gpx1000 committed:
ios simulator isn't supported yet.

gpx1000 committed:
ios simulator isn't supported yet.

gpx1000 committed:
allow updated FindVulkan.cmake to be used.
Fix the paths such that environment variable expects VULKAN_SDK to end in macOS.

gpx1000 committed:
fix ci

gpx1000 committed:
fix CMake less than 3.29 and MoltenVK.framework.

gpx1000 committed:
Force CMake to only have one place to search in CMAKE 3.29 and above for MoltenVK.

gpx1000 committed:
Force CMake to only have one place to search in CMAKE 3.29 and above for MoltenVK.

richard-lunarg commented:
I've tried this locally. As of today, it builds with the current macOS VulkanSDK, and the upcoming 1.3.280.1 release.

richard-lunarg reviewed (approved):
I've tried this PR locally. It builds with the latest cmake and Xcode versions for me and I can deploy to an iOS device(s).

jeroenbakker-atmind commented:
I tested compiling this PR with VulkanSDK 1.3.275.0 to ensure old workflows still work. Validated both the CMake and Xcode path. Both worked as expected. I don't have an iOS (or simulator) currently installed, but that has already been validated by others.
